OpenACC backend

jkelling opened the pull request:
~*Builds on https://github.com/alpaka-group/alpaka/pull/1126  and will be rebased accordingly.*~

This PR adds an OpenACC offloading backend.

## CMake

* Adds cmake option `ALPAKA_ACC_ANY_BT_OACC_ENABLE`.
  * Compiler flags enabling OpenMP and OpenACC can be mutually exclusive, thus this backend cannot be enabled at the same time as any of the OpenMP backends. A CMake warning for this case is in place.

## CI

* Adds CI job linux_gcc-10_release-oacc which only enables this backend and `ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLE`
* All tests with multiple compilation units are disabled for OpenACC+GCC because of a bug where GCC messes up symbols of target regions in some way.
* The test math is disabled for OpenACC+GCC because GCC does not allow types with `static constexpr` members to be mapped.
* In terms of compiler support: Most examples and tests can be compiled for tesla devices using NVHPC 20.7 (others trigger ICEs). Most compilable tests also succeed on GPU.

## Backend Properties

* Adds `AccOacc`, analogous to `AccOmp5`
* Adds `DevOacc` to handle OpenACC device ids, type and `acc_set_device_num`.
* Adds `BufOacc` for OpenACC device allocations.
* Uses `QueueGenericThreads*<DevOacc>`.
* ~Adds `CtxThreadOacc` to represent the context of a thread/worker.~ A per-thread instance [of `AccOacc`] is required because there is no intrinsic way in OpenACC to get a worker id and thus a loop counter ist used.
* Adds `CtxBlockOacc` to represent block-shared state, like shared memory and infrastructure for synchronization.

~While `AccOacc` is a handle for the backend on the host side, it doe not implement all concepts other `Acc*` type do. Concepts for traits which can only be used in device code, such as `getWorkDiv` or `getIdx` are only implemented by `CtxThreadOacc` which derives from `AccOacc`.~ Each instance of ~`CtxThreadOacc`~`AccOacc` holds a pointer to a block-shared instance of `CtxBlockOacc`.

## Quirks

* OpenACC does not support intra-block(gang) synchronization. `SyncBlockThreads<>::syncBlockThreads` had to be implemented using atomics: https://github.com/jkelling/alpaka/blob/openacc/include/alpaka/ctx/block/CtxBlockOacc.hpp#L114
* `ALPAKA_OFFLOAD_MAX_BLOCK_SIZE` is used to set the max number of threads per block (workers per gang). The runtime may choose a lower number than requested without warning. There is no way of determining maximum number of workers per gang the runtime will allow at compile time *or run time*. If the runtime runs fewer workers than requested, this will lead `SyncBlockThreads<>::syncBlockThreads` to dead lock. Thus the user setting the correct value their runtime will allow is crucial. This should probably be put into the documentation, but I am not familiar with Alpaka's documentation process. It is documented in the code.

jkelling committed:
Add OpenACC backend

OpenACC: BufOacc, Copy

OpenACC: sort GetIdx traits and AccCpuOaccWorker

OpenACC: sort atomics; Add atomic-based gang barrier

TaskCopyOacc: Add makeTaskCopyOacc, to deduce function object type

OpenACC: Add SyncBlockThreads

OpenACC::SyncBlockThreads: Fix atomic capture

for PGI, but GCC eats it too

OpenACC::atomics: use acc capture {} form

OpenACC: Fix block/shared/st/BlockSharedMemStOacc.hpp

AccCpuOacc: Fix comments

penACC Rm AccCpuOaccWorker, Add Ctx{Block,Thread}Oacc

OpenACC: Fix Acc workdiv order

OpenACC: Rename AccCpuOacc -> AccOacc

Fix PltfOacc,DevOacc GetDev.., ctor

buf/oacc/Copy: Fix template<> typename -> class

BlockSyncBarrierOacc: Fix include sync Traits

CtxBlockOacc:Fix missing forward declration of TaskKernelOacc

AccOacc: Use WorkDivMembers

OpenACC: excludes/guards for tests

OpenACC: Fixes to AccOacc structure/Traits

OpenACC: Add missing atomics

Min, Max, Inc, Dec, Cas

test/unit/atomic: Fix AtomicTestKernel

Add template<TTAcc>

buf/oacc/Copy: Add <Oacc, Oacc> for OpenACC<2.5

AccOacc: Use CtxThreadAcc instead

Add CreateStaticDevMemView<dev::DevOacc>

OpenACC: Add SyncBlockThreadsPredicate<CtxThreadOacc>

OpenACC: BlockSharedMem*Oacc: use array<size> for alignment

OpenAcc::atomic: use ref instead of old in expressions

... to silence GCC warning -Wuninitialized, -Wmaybe-uninitilized

BlockSharedMemDynOacc: Fix size_t -> std::size_t

Fix debug output in alpaka//oacc/Copy.hpp

Fix some omp4 leftovers in oacc backend

Add README_OACC.md

oacc: Rm acc MemSetKernel: use shared one

mem/buf/oacc/Set.hoo: clean-up omp4 leftovers

README_OACC: Add details on examples

oacc/Copy: Add GCC10.1 ! __GNUC__ check

to guard acc_memcpy_device

BlockSharedMemStOacc: Fix modulus in alignPitch

Oacc: use BlockSharedMem*Member

DevOacc: Add concept Dev

ExampleDefaultAcc: Add Oacc, Update Omp4->Omp5

DevOacc: Fix missing ,

README_OACC: Updae

Add standalone/AnyOacc.hpp

AtomicOaccBuiltIn: disable atomic exch for PGI TPR28628

Add GetWarpSize<DevOacc> stub

TaskKernelOacc: Add ALPAKA_DEBUG gurads to prints

TaskKernelOacc: Add copyin(), default(present); Rm teamnum loop

Co-authored-by: Mat Colgrove <mcolgrove@nvidia.com>

TaskKernelOacc: Rm OpenMP 4 relics; Fix indent

TaskKernelOacc: Use ALPAKA_DECAY_T

GetAccDevProps<AccOacc<>>: Add staticAllocBytes()

Oacc: Add DevOacc::makeCurrent print (ALPAKA_FULL_DEBUG)

Add ALPAKA_FULL_DEBUG guard for BufOacc alloc print

BlockSyncBarrierOacc::m_syncCounter: Add comment about NVHPC warning

Oacc: disable tests separableCompilation, atomic

AccOacc: Add base IntrinsicFallback

AccOacc: Add base WarpSingleThread

Oacc: getMem<ctx::CtxThreadOacc<...>>: Add GNUC -Wcast-align ignore

Disable tests with multiple object files for Oacc+GCC10

IdxGbOaccBuiltIn: use IdxGbLinear as BlockShared

AccOacc: Fix gang limit in CI

CI: Add Oacc

update README_OACC: new cmake defaults

Oacc: re-enable multi-object test for GCC 11: linker bug fxed

DevOacc: replace CUDA RT left-overs

TaskKernelOacc:::operator(): Add explicit empty grid -> NOP

buf/oacc/Set: -> return empty grid on empty range

DevOacc: Add implement get{,Free}MemBytes

update README_OACC; add limitations

QueueOacc*Blocking: Rm superflous includes

Fix CtxBlockOacc::ctor blockSharedMemDynSizeBytes type

.. to adjust to https://github.com/alpaka-group/alpaka/pull/1141

mcolg committed:
CtxBlockOacc: SyncBlockThreads: Add workaround for NVHPC issue

bernhardmgruber commented on include/alpaka/atomic/AtomicOaccBuiltIn.hpp:
```diff
@@ -0,0 +1,387 @@
+
+#pragma once
+
+#ifdef _OPENACC
```
Shouldn't this be `ALPAKA_ACC_ANY_BT_OACC_ENABLED`?

jkelling replied:
No, the pragma-containing files are also guarded this way in the OpenMP backends: https://github.com/alpaka-group/alpaka/blob/develop/include/alpaka/atomic/AtomicOmpBuiltIn.hpp#L12

jkelling replied:
Ok ... maybe this is because the OMP atomics are shared between multiple backends, this is not the case for OpenACC atomic, i.e. it could be guarded by `ALPAKA_ACC_ANY_BT_OACC_ENABLED`, indeed.

jkelling replied:
Which way of consistency is better?

psychocoderHPC replied:
~~I think the current code is fine else you need an include before the guard.~~

psychocoderHPC replied:
Ohh I am wrong `ALPAKA_ACC_ANY_BT_OACC_ENABLED` is coming from CMake and not from a header file. 
Then I would say `ALPAKA_ACC_ANY_BT_OACC_ENABLED` is the right way to be consistent with the other OpenACC files.

bernhardmgruber commented on include/alpaka/block/sync/BlockSyncBarrierOacc.hpp:
```diff
@@ -0,0 +1,70 @@
+
+#pragma once
+
+#ifdef _OPENACC
```
See above

bernhardmgruber commented on include/alpaka/kernel/TaskKernelOacc.hpp:
```diff
@@ -0,0 +1,345 @@
+                            for(TIdx w = 0; w < blockThreadCount; ++w)
+                            {
+                                // blockThreadIdx[0] = w;
+                                ctx::CtxThreadOacc<TDim, TIdx> acc(
+                                    w,
+                                    blockShared);
```
Is there a serious reason for these linebreaks?

jkelling replied:
Style: Either all-in-one-line or one-per-line.

jkelling replied:
Same format as used in the OpenMP backends in this place: https://github.com/alpaka-group/alpaka/blob/develop/include/alpaka/kernel/TaskKernelCpuOmp2Blocks.hpp#L214 .

bernhardmgruber commented on include/alpaka/kernel/TaskKernelOacc.hpp:
```diff
@@ -0,0 +1,345 @@
+                                meta::apply(
+                                    [kernelFnObj, &acc](typename std::decay<TArgs>::type const & ... args)
+                                    {
+                                        kernelFnObj(
+                                                acc,
+                                                args...);
```
why the line breaks?

bernhardmgruber commented on include/alpaka/queue/QueueOaccBlocking.hpp:
```diff
@@ -0,0 +1,42 @@
+    #error If ALPAKA_ACC_ANY_BT_OACC_ENABLED is set, the compiler has to support OpenACC xx or higher!
+#endif
+
+#include <alpaka/dev/DevOacc.hpp>
+
+#include <alpaka/dev/Traits.hpp>
+#include <alpaka/event/Traits.hpp>
+#include <alpaka/queue/Traits.hpp>
+#include <alpaka/wait/Traits.hpp>
+
+#include <alpaka/queue/QueueGenericThreadsBlocking.hpp>
+
+#include <stdexcept>
+#include <memory>
+#include <functional>
+#include <mutex>
+#include <condition_variable>
+#include <thread>
+
+namespace alpaka
+{
+    namespace queue
+    {
+        using QueueOaccBlocking = QueueGenericThreadsBlocking<dev::DevOacc>;
+    }
+}
```
Feels like too many includes for this one alias definition ;)

jkelling replied:
True.

bernhardmgruber commented on include/alpaka/queue/QueueOaccNonBlocking.hpp:
```diff
@@ -0,0 +1,42 @@
+    #error If ALPAKA_ACC_ANY_BT_OACC_ENABLED is set, the compiler has to support OpenACC xx or higher!
+#endif
+
+#include <alpaka/dev/DevOacc.hpp>
+
+#include <alpaka/dev/Traits.hpp>
+#include <alpaka/event/Traits.hpp>
+#include <alpaka/queue/Traits.hpp>
+#include <alpaka/wait/Traits.hpp>
+
+#include <alpaka/queue/QueueGenericThreadsNonBlocking.hpp>
+
+#include <stdexcept>
+#include <memory>
+#include <functional>
+#include <mutex>
+#include <condition_variable>
+#include <thread>
+
+namespace alpaka
+{
+    namespace queue
+    {
+        using QueueOaccNonBlocking = QueueGenericThreadsNonBlocking<dev::DevOacc>;
+    }
+}
```
Here as well

bernhardmgruber committed:
apply interface changes after recent namespace refactoring

jkelling committed:
OpenACC backend PR fixes

Fix ExampleDefaultAcc for OAcc -> alpaka::CtxThreadOacc<TDim,TIdx>

Fix TestAccs: Rm double AccOacc

Oacc Atomic,Sync: Fix guards

IdxBtOacc: rename, fix comment

Oacc: clean-up comments

Oacc: Buf,Copy: clean-up code

Oacc: Fix ci.yml

Oacc: merge CtxThreadOacc into AccOacc

AccOacc: Add final

dtor can never be virtual because of use in device code

Oacc::Copy: remove cast on compare to 0.

oacc: Remove /* virtual */ dtor comments

Oacc: rename IdxGbOacc

Oacc: clean-up code, comments

AtomicTestKernel::operator(): Rm TTAcc

README_OACC: Rm -fopenacc/-acc (set by CMake)

Mv IdxBtOacc -> IdxBtLinear

Oacc: Rm *::BlockShared

Oacc/Set: Add debug guard around print

BufOacc: Fix size DBG output

BufOmp5: Fix size DBG output

check OpenACC 2.0, error below

jkelling commented:
@psychocoderHPC can you please have a look at this PR? Many issues have already been fixed when they came up in https://github.com/alpaka-group/alpaka/pull/1126 , it does however have some fundamentally different aspects to it.

I am currently thinking about merging `CtxThreadOacc` into `AccOacc`. My initial consideration leading to separating them, was that `CtxThreadOacc` must only be instanceciated in device code and must be thread-private, but if i am not mistaken, `Acc`s are also never instanceciated outside of device code, and are specifically only created in the respective `TaskKernel*::operator()`. Is this correct? `CtxBlockOacc` could be moved into the namespace ~~`detail::oacc`~~ `oacc::detail`, as user-code should not need to interact with it directly.

psychocoderHPC commented:
> @psychocoderHPC can you please have a look at this PR? Many issues have already been fixed when they came up in #1126 , it does however have some fundamentally different aspects to it.
> 
> I am currently thinking about merging `CtxThreadOacc` into `AccOacc`. My initial consideration leading to separating them, was that `CtxThreadOacc` must only be instanceciated in device code and must be thread-private, but if i am not mistaken, `Acc`s are also never instanceciated outside of device code, and are specifically only created in the respective `TaskKernel*::operator()`. Is this correct? `CtxBlockOacc` could be moved into the namespace ~`detail::oacc`~ `oacc::detail`, as user-code should not need to interact with it directly.

Yes `Acc`'s will never be instantiated outside of kernels, so living only on the device.

BenjaminW3 commented on include/alpaka/idx/bt/IdxBtOaccBuiltIn.hpp:
```diff
@@ -0,0 +1,128 @@
+    namespace bt
+    {
+        //#############################################################################
+        //! The CUDA accelerator ND index provider.
```
CUDA?

BenjaminW3 commented on include/alpaka/idx/bt/IdxBtOaccBuiltIn.hpp:
```diff
@@ -0,0 +1,128 @@
+        template<
+            typename TDim,
+            typename TIdx>
+        class IdxBtOaccBuiltIn : public concepts::Implements<ConceptIdxBt, IdxBtOaccBuiltIn<TDim, TIdx>>
```
Why is this called `BuiltIn`? I do not see it accessing a compiler built-in. It only holds the index as a member.

jkelling replied:
Copy from OMP, fixed.

BenjaminW3 commented on include/alpaka/kernel/TaskKernelOacc.hpp:
```diff
@@ -0,0 +1,313 @@
+                        #pragma acc loop worker
+                        for(TIdx w = 0; w < blockThreadCount; ++w)
+                        {
+                            // blockThreadIdx[0] = w;
```
What is this comment telling me?

BenjaminW3 commented on include/alpaka/mem/buf/BufOacc.hpp:
```diff
@@ -0,0 +1,652 @@
+                    BufOaccImpl(BufOaccImpl &&) = default;
+                BufOaccImpl& operator=(const BufOaccImpl&) = delete;
+                BufOaccImpl& operator=(BufOaccImpl&&) = default;
+                    ~BufOaccImpl()
```
Why is there a difference in indentation here?

jkelling replied:
Ctors. Wrong style, fixed.

BenjaminW3 commented on include/alpaka/mem/buf/BufOacc.hpp:
```diff
@@ -0,0 +1,652 @@
+        {}
+
+            BufOacc(const BufOacc&) = default;
+            BufOacc(BufOacc &&) = default;
```
And here?

BenjaminW3 commented on include/alpaka/mem/buf/BufOacc.hpp:
```diff
@@ -0,0 +1,652 @@
+
+                std::size_t size = static_cast<std::size_t>(extent[0]*static_cast<TIdx>(sizeof(TElem)));
+                for (unsigned int a = 1u; a < static_cast<unsigned int>(TDim::value); ++a)
+                    size *= static_cast<std::size_t>(extent[a]);
```
I would prefer braces even for such short loops.

BenjaminW3 replied:
But I would even more prefer the usage of `const auto size = static_cast<std::size_t>(extent.prod()) * sizeof(TElem)`

BenjaminW3 commented on include/alpaka/mem/buf/oacc/Copy.hpp:
```diff
@@ -0,0 +1,395 @@
+                                    reinterpret_cast<void *>(
+                                        this->m_dstMemNative + (castVec<DstSize>(idx) * dstPitchBytesWithoutOutmost).foldrAll(std::plus<DstSize>())),
+                                    const_cast<void*>(reinterpret_cast<const void *>(
+                                        this->m_srcMemNative + (castVec<SrcSize>(idx) * srcPitchBytesWithoutOutmost).foldrAll(std::plus<SrcSize>()))),
```
There is a shortcut for `foldrAll(plus)` called `sum()`.

jkelling replied:
Did not notice this member, thanks.

BenjaminW3 commented on include/alpaka/mem/buf/oacc/Copy.hpp:
```diff
@@ -0,0 +1,395 @@
+#else
+                            // acc_memcpy_device is only available since OpenACC2.5
+                            // , but we want the tests to compile anyway
+                            [](void* dst, void* src, std::size_t size){
```
why is the memory pointed to by src not const? This makes the hacky `const_cast` above necessary.

jkelling replied:
Because it is not const in the OpenACC API. The `const_cast` must happen somewhere.

BenjaminW3 commented on include/alpaka/mem/buf/oacc/Set.hpp:
```diff
@@ -0,0 +1,107 @@
+                byteExtent[TDim::value-1] *= static_cast<Idx>(sizeof(Elem<TView>));
+                constexpr auto lastDim = TDim::value - 1;
+
+                if(pitch[0] <= 0)
```
I would prefer braces here.

bernhardmgruber replied:
I would not :D But I guess @j-stephan will come up with that style guide discussion at some point for clang-tidy.

jkelling replied:
I am fine with braces, will fix.

BenjaminW3 commented on include/alpaka/pltf/PltfOacc.hpp:
```diff
@@ -0,0 +1,94 @@
+            PltfOacc>
+        {
+            //-----------------------------------------------------------------------------
+            //! \param devIdx device id, less than GetDevCount, will be set to omp_get_initial_device() if < 0
```
omp_get_initial_device?

BenjaminW3 commented on test/unit/atomic/src/AtomicTest.cpp:
```diff
@@ -425,8 +425,10 @@ class AtomicTestKernel
+        typename TTAcc>
     ALPAKA_FN_ACC auto operator()(
-        TAcc const & acc,
+        TTAcc const & acc,
```
Why?

jkelling replied:
Kernel `operator()` must itself be a template with respect to `acc`, so the backend can pass the thread context.

The class already has `TAcc` as template parameter. This patch adds a template parameter `TTAcc` for the `operator()`, named this way to not trigger `-Wshadow`.

bernhardmgruber replied:
I do not get it. If `operator()` needs to have the template parameter, then `AtomicTestKernel` should not have it. Please remove it from the class then. Or does the class support to be instantiated on two different accelerators?

jkelling replied:
All tests are explicitly templated on the accelerator. I did not know what implications removing the template here, because this could go into catch2. I can have a try.

jkelling replied:
The template param `TAcc` cannot be removed from the class `AtomicTestKernel` because it is specialized on this param for the HIP backend. However, technically, after merging `CtxThreadOacc` and `AccOacc` the template param on the `operator()` is no longer required, but not having it does not appear clean. Still, I will remove it from this PR.

BenjaminW3 commented on .github/workflows/ci.yml:
```diff
@@ -242,6 +243,9 @@ jobs:
           env: {CXX: g++,     CC: gcc,    ALPAKA_CI_GCC_VER: 10,       ALPAKA_CI_STDLIB: libstdc++, CMAKE_BUILD_TYPE: Release, ALPAKA_CI_BOOST_BRANCH: boost-1.74.0, ALPAKA_CI_CMAKE_VER: 3.17.3, OMP_NUM_THREADS: 2, ALPAKA_CI_DOCKER_BASE_IMAGE_NAME: "ubuntu:20.04"}
+        - name: linux_gcc-10_release_oacc
+          os: ubuntu-latest
+          env: {CXX: g++,     CC: gcc,    ALPAKA_CI_GCC_VER: 10,       ALPAKA_CI_STDLIB: libstdc++, CMAKE_BUILD_TYPE: Release, ALPAKA_CI_BOOST_BRANCH: boost-1.74.0, ALPAKA_CI_CMAKE_VER: 3.17.3, OMP_NUM_THREADS: 2, ALPAKA_CI_DOCKER_BASE_IMAGE_NAME: "ubuntu:20.04", CMAKE_CXX_FLAGS: "-fopenmp -foffload=disable", ALPAKA_ACC_ANY_BT_OACC_ENABLE: ON, ALPAKA_OFFLOAD_MAX_BLOCK_SIZE: 1, ACC_DEVICE_TYPE: "h ...
```
Why do I need `-fopenmp` for openacc?

jkelling replied:
Fixed. (also no `-fopenacc` required, because it is added by cmake)

bernhardmgruber commented on README_OACC.md:
````diff
@@ -0,0 +1,89 @@
+
+```bash
+make helloWorld
+./examples/helloWorld/helloWorld
+```
+The output should end with something like
+```
+[z:3, y:7, x:15][linear:511] Hello World
+```
+Numbers can vary when teams are executed in parallel: 512 teams, with one worker
+each are started in a 3d grid. Each worker reports its grid coordinates and linear
+index.
+
+|compiler|compile status|target|run status|
+|---|---|---|---|
+|GCC 10| ok|x86|ok|
+|NVHPC 20.7| ok|tesla|ok|
+
+### vectorAdd
+
+```bash
+make vectorAdd
+./examples/vectorAdd/vectorAdd
+```
+The output should end with
+```
+Execution results correct!
+```
+
+|compiler|compile status|target|run status|
+|---|---|---|---|
+|GCC 10(dev)| ok|x86|ok|
+|NVHPC 20.7| ok|tesla|ok|
+
+## Building and Running all tests
+
+```bash
+make
+ctest
+```
````
Why is this documentation in an extra README file? I know we already have README_OMP (same question applies here). But I still wonder, if this shoud be part of the main README file, or moved somewhere else inside the documentation. And if this extra README file is fine, why is there no README for CUDA or HIP?

jkelling replied:
This file has two purposes:
1. Give testers, including those from compiler vendors, a quick intro into how to test the OpenACC backend specifically.
2. To list and somewhat keep track of, nown compiler issues with this backend.

Here, like with OpenMP target, compiler support is severely lacking. For OpenACC there is no compiler with which this actually works completely and there are no workarounds for some issues. It only, hopefully, works in theory. Once there is a compiler which could be considered production ready, this readme can be removed.

bernhardmgruber commented on cmake/alpakaCommon.cmake:
```diff
@@ -55,6 +55,7 @@ option(ALPAKA_ACC_CPU_B_TBB_T_SEQ_ENABLE "Enable the TBB CPU grid block back-end
 option(ALPAKA_ACC_CPU_B_OMP2_T_SEQ_ENABLE "Enable the OpenMP 2.0 CPU grid block back-end" OFF)
 option(ALPAKA_ACC_CPU_B_SEQ_T_OMP2_ENABLE "Enable the OpenMP 2.0 CPU block thread back-end" OFF)
 option(ALPAKA_ACC_ANY_BT_OMP5_ENABLE "Enable the OpenMP 5.0 CPU block and block thread back-end" OFF)
+option(ALPAKA_ACC_ANY_BT_OACC_ENABLE "Enable the OpenACC block and block thread back-end" OFF)
```
A `block and block thread` sounds weird. Why not just `Enable the OpenACC back-end`?

jkelling replied:
Consistency with the OpenMP backends. I think these name allude to an early idea that one could combine backends for thread and block level, so backends always say, that they handle both block and thread-level, not allowing combination with others. Changing this would require a separate discussion and should be done for all backends.

bernhardmgruber replied:
Ok!

bernhardmgruber commented on cmake/alpakaCommon.cmake:
```diff
@@ -258,6 +266,14 @@ if(ALPAKA_ACC_CPU_B_OMP2_T_SEQ_ENABLE OR ALPAKA_ACC_CPU_B_SEQ_T_OMP2_ENABLE OR A
+if(ALPAKA_ACC_ANY_BT_OACC_ENABLE)
+   find_package(OpenACC)
+   if(OpenACC_CXX_FOUND)
+      target_compile_options(alpaka INTERFACE ${OpenACC_CXX_OPTIONS})
+      target_link_options(alpaka INTERFACE ${OpenACC_CXX_OPTIONS})
```
According to https://cmake.org/cmake/help/v3.19/module/FindOpenACC.html it seems CMake offers imported targets (since versio 3.16) which should be preferred. So something like `target_link_library(alpaka INTERFACE OpenACC::OpenACC_CXX)` should be sufficient.

jkelling replied:
```cmake
target_link_options(alpaka INTERFACE OpenACC::OpenACC_CXX)
```
yields (with cmake 3.18.1)
```
$ g++ -foffload=disable -Wall OpenACC::OpenACC_CXX CMakeFiles/vectorAdd.dir/src/vectorAdd.cpp.o -o vectorAdd  -lpthread /usr/lib/x86_64-linux-gnu/librt.so
g++: error: OpenACC::OpenACC_CXX: No such file or directory
```
There is no use of this for OpenMP either in Alpaka, so no working example. Do you know the exact syntax for this?

bernhardmgruber replied:
Hmm. Then leave it as is for now. I just looked at the CMake docs and thought it should work with imported targets.

bernhardmgruber commented on include/alpaka/acc/AccOacc.hpp:
```diff
@@ -0,0 +1,266 @@
+    template<
+        typename TDim,
+        typename TIdx>
+    class AccOacc :
```
Any reason this is not called `AccOpenACC`? Is this too long? `AccOacc` looks a bit cryptic to me.

jkelling replied:
Because the OpenMP backends are also called `Acc...Omp...`, stuck with the same pattern.

bernhardmgruber commented on include/alpaka/acc/AccOacc.hpp:
```diff
@@ -0,0 +1,266 @@
+        //-----------------------------------------------------------------------------
+        auto operator=(AccOacc &&) -> AccOacc & = delete;
+        //-----------------------------------------------------------------------------
+        /*virtual*/ ~AccOacc() = default;
```
Why is there a commented out `virtual`? If the accelerator should not be inherited from, then why is it not final?

jkelling replied:
The `virtual` here is pointless either way (left over from OpenMP CPU): This is used on GPU, where not virtual functions are allowed. Making it final.

bernhardmgruber commented on include/alpaka/acc/AccOacc.hpp:
```diff
@@ -0,0 +1,266 @@
+                auto const gridBlockCountMax(alpaka::core::clipCast<TIdx>(oaccMaxGangNum));
+#endif
+                return {
+                    // m_multiProcessorCount
+                    static_cast<TIdx>(gridBlockCountMax),
```
Sidenote: In C++20 we have designated initializers and you could write: `.m_multiProcessorCount = static_cast<TIdx>(gridBlockCountMax)`. The comments may be nice, but they don't protect you if the constructed struct changes.

bernhardmgruber commented on include/alpaka/acc/AccOacc.hpp:
```diff
@@ -0,0 +1,266 @@
+            ALPAKA_FN_HOST static auto getAccName()
+            -> std::string
+            {
+                return "AccOacc<" + std::to_string(TDim::value) + "," + typeid(TIdx).name() + ">";
```
`typeid(TIdx).name()` will often result in cryptic names. `boost::core::demangle` improves this a fair bit!

BenjaminW3 replied:
We are still in the progress of trying to get rid of the boost dependency so we should not add a new one ;-)

bernhardmgruber replied:
Fair enough!

bernhardmgruber commented on include/alpaka/acc/AccOacc.hpp:
```diff
@@ -0,0 +1,266 @@
+        };
+
+        //#############################################################################
+        //! The CPU OpenACC accelerator device type trait specialization.
```
I fail to understand here and in the following comments, what the accelerator has to do with the `CPU`?

bernhardmgruber commented on include/alpaka/alpaka.hpp:
```diff
@@ -25,6 +25,8 @@
 #include <alpaka/acc/AccCpuOmp2Blocks.hpp>
 #include <alpaka/acc/AccCpuOmp2Threads.hpp>
 #include <alpaka/acc/AccOmp5.hpp>
+// #include <alpaka/acc/AccOacc.hpp>
```
If the header is not needed, please remove it.

bernhardmgruber replied:
@jkelling

bernhardmgruber replied:
@jkelling

bernhardmgruber commented on include/alpaka/atomic/AtomicOaccBuiltIn.hpp:
```diff
@@ -0,0 +1,384 @@
+            -> T
+            {
+                T old;
+                auto & ref(*addr);
+                // atomically update ref, but capture the original value in old
+                #pragma acc atomic capture
+                {
+                    old = ref;
+                    ref += value;
+                }
```
I fail to understand why that extra `ref` variable helps with anything. Why is it there?

psychocoderHPC replied:
I assume the standard defines that the region which defines an atomic must have s special style. Casts, derefs or anything else is not allowed. The code is used for OMP3.X+ too.

bernhardmgruber replied:
Hmm. Maybe. But it seems weird that pointers would not be supported but references would. The latter are basically syntatic sugar for pointers.

psychocoderHPC replied:
I agree, you can have a look https://www.openmp.org/spec-html/5.0/openmpsu95.html to see what I mean.

jkelling replied:
+1

With the currently state of things, I am not going to argue with compilers on this one.

bernhardmgruber commented on include/alpaka/ctx/block/CtxBlockOacc.hpp:
```diff
@@ -0,0 +1,302 @@
+// The change is sematically equivlent.
+// However, this should work per the OpenACC standard, but appears to be compiler
+// issue causing a runtime error.  The error was seen the 20.7 release
+// of the NVIDIA HPC Compiler but may be corrected in future releases.
```
Why is this not indented?

jkelling replied:
Sorry, missed this one. Now fixed in https://github.com/alpaka-group/alpaka/pull/1127/commits/d7936f807e55de28bf6f858cd2b84790e24eadd8 .

bernhardmgruber commented on include/alpaka/dev/DevOacc.hpp:
```diff
@@ -0,0 +1,343 @@
+                    m_deviceType(::acc_get_device_type()),
+                    m_iDevice(iDevice)
+                {}
+                // DevOaccImpl(int iDevice) : m_iDevice(iDevice) {}
```
Dead code? If yes, please remove.

bernhardmgruber replied:
@jkelling

bernhardmgruber replied:
@jkelling

bernhardmgruber commented on include/alpaka/dev/DevOacc.hpp:
```diff
@@ -0,0 +1,343 @@
+                    std::lock_guard<std::mutex> lk(m_Mutex);
+
+                    // Register this queue on the device.
+                    m_queues.push_back(spQueue);
```
Unnecessary copy of `spQueue`, please add a `std::move`.

bernhardmgruber replied:
@jkelling

bernhardmgruber replied:
@jkelling

bernhardmgruber commented on include/alpaka/dev/DevOacc.hpp:
```diff
@@ -0,0 +1,343 @@
+        //-----------------------------------------------------------------------------
+        auto operator=(DevOacc &&) -> DevOacc & = default;
+        //-----------------------------------------------------------------------------
+        ALPAKA_FN_HOST auto operator==(DevOacc const & rhs) const
+        -> bool
+        {
+            return m_spDevOaccImpl->iDevice() == rhs.m_spDevOaccImpl->iDevice();
+        }
+        //-----------------------------------------------------------------------------
+        ALPAKA_FN_HOST auto operator!=(DevOacc const & rhs) const
+        -> bool
+        {
+            return !((*this) == rhs);
+        }
```
Prefer defining relational operators as hidden friends.

BenjaminW3 replied:
I think it is done this way for consistency with the other devices which do the same. We should not diverge here but do it consistently. As this is already a large PR, changing the other impls should not be added to this PR so the change should an independent PR.

bernhardmgruber replied:
I created issue #1201 for this.

bernhardmgruber commented on include/alpaka/dev/DevOacc.hpp:
```diff
@@ -0,0 +1,343 @@
+                DevOacc const &)
+            -> std::string
+            {
+                return std::string("OpenACC target");
```
The extra construction of a string is not necessary. `return "OpenACC target"` should suffice.

bernhardmgruber replied:
@jkelling

jkelling replied:
I know, that there is an implicit cast, but alpaka likes to be very explicit. This is done in the same way in e.g. the CUDA backend: https://github.com/alpaka-group/alpaka/blob/develop/include/alpaka/dev/DevUniformCudaHipRt.hpp#L122

bernhardmgruber commented on include/alpaka/dev/DevOacc.hpp:
```diff
@@ -0,0 +1,343 @@
+        {
+            //-----------------------------------------------------------------------------
+            ALPAKA_FN_HOST static auto getWarpSize(
+                DevOacc const & dev)
+            -> std::size_t
+            {
+                alpaka::ignore_unused(dev);
```
If `dev` is not used, then by declare it in the first place? You can just have `DevOacc const&)` in the parameter list.

BenjaminW3 replied:
But then you would lose the parameter name completely which may not be desired because it might contain usage hints or because we may want to add doxygen documentation for the parameter. Both `alpaka::ignore_unused` as well as removing the name are not the right solution. The only correct solution is the c++17 `maybe_unused` attribute which we can not use yet. For now `alpaka::ignore_unused` is the consisten way.

bernhardmgruber replied:
I know these reasons, but there is neither a doxygen comment nor does the name `dev` help here ;) But we can leave it as is.

bernhardmgruber commented on include/alpaka/kernel/TaskKernelOacc.hpp:
```diff
@@ -0,0 +1,313 @@
+
+                        // Execute the threads in parallel.
+
+                        // Parallel execution of the threads in a block is required because when syncBlockThreads is called all of them have to be done with their work up to this line.
```
That line seems to long.

BenjaminW3 replied:
clang-format will care about this ;-)

jkelling replied:
fixed, although the exact length was no decided on yet.

bernhardmgruber commented on include/alpaka/mem/buf/BufOacc.hpp:
```diff
@@ -0,0 +1,652 @@
+        oacc::detail::BufOaccImpl<TElem, TDim, TIdx>& operator*() {return *m_spBufImpl;}
+        const oacc::detail::BufOaccImpl<TElem, TDim, TIdx>& operator*() const {return *m_spBufImpl;}
+
+        inline const Vec<TDim, TIdx>& extentElements() const {return m_spBufImpl->m_extentElements;}
```
Class member functions are automatically inline. No need for the `inline` keyword.

bernhardmgruber replied:
@jkelling

bernhardmgruber commented on include/alpaka/mem/buf/BufOacc.hpp:
```diff
@@ -0,0 +1,652 @@
+        BufOacc& operator=(const BufOacc&) = default;
+        BufOacc& operator=(BufOacc&&) = default;
+
+        oacc::detail::BufOaccImpl<TElem, TDim, TIdx>& operator*() {return *m_spBufImpl;}
+        const oacc::detail::BufOaccImpl<TElem, TDim, TIdx>& operator*() const {return *m_spBufImpl;}
```
Using `operator*` seems like a weird thing to do. In what context will this be used?

bernhardmgruber replied:
@jkelling

jkelling replied:
This is basically a getter to access the `Impl` instance as a reference. In a way, this is a pointer type as it is implicitly shared.

The way this is implemented here is the same as in the `BufOmp5`. It is different from `BufCpu` only in that the impl pointer is private which I find to be cleaner than having it public. These operators are only used by trait specializations and for these it felt more explicit to do the `(*buf).`-thing, then use `buf->` when I die this for `BufOmp5`.

A clean implementation of an implicitly shared type would not expose the impl pointer at all and instead wrap all the impl's functionality with member functions or have the traits as friends (the former being more clean.). I think there was a discussion on this before. This is not done in the clean way in alpaka currently, probably because of the amounts of boiler-plate involved.

I suggest to leave this as-is for now and maybe open an against both `BufOmp5` and `BufOacc` (maybe also `BufCpu`) to clean this up.

bernhardmgruber replied:
I feel like we should revisit the syntax at some point. `operator*` does not feel like the right choice here. Especially, if a corresponding `operator->` is not provided. But then I also do not understand the shared buffer implementation well enough. Leave it as is for now.

jkelling replied:
Opened issue https://github.com/alpaka-group/alpaka/issues/1207 .

bernhardmgruber commented on include/alpaka/mem/buf/BufOacc.hpp:
```diff
@@ -0,0 +1,652 @@
+                BufOacc<TElem, TDim, TIdx> const & buf)
+            -> DevOacc
+            {
+                return (*buf).m_dev;
```
That looks very strange to me, since `buf` is a reference. So `BufOacc<TElem, TDim, TIdx>` models some kind of pointer type?

bernhardmgruber replied:
@jkelling

bernhardmgruber commented on include/alpaka/mem/buf/BufOacc.hpp:
```diff
@@ -0,0 +1,652 @@
+                BufOacc<TElem, TDim, TIdx> const & buf)
+            -> TElem const *
+            {
+                return (*buf).m_pMem;
```
Why not? `buf->m_pMem`? If you provide `operator*` and `operator->` should be there as well.

bernhardmgruber replied:
@jkelling

bernhardmgruber commented on include/alpaka/mem/buf/oacc/Copy.hpp:
```diff
@@ -0,0 +1,395 @@
+#if ALPAKA_DEBUG >= ALPAKA_DEBUG_FULL
+                    this->printDebug();
+#endif
+                    if(static_cast<std::size_t>(this->m_extent.prod()) != 0u)
```
`if(this->m_extent.prod() != 0)` should be enough.

jkelling replied:
Sometimes (in some tests or examples), extents are signed, causing a warning without the cast.

bernhardmgruber replied:
That's why I would use `0` and not `0u`, as I wrote.

jkelling replied:
Aha, I did not notice that.

Would this not also raise a sign compare warning in the opposite case, i.e. when extent is unsigned?

bernhardmgruber replied:
It should not, because the compiler can see that the int literal 0 does not change it's value when promoted to an unsigned integral. But you are free to proof me wrong! ;)

jkelling replied:
First I would need to figure out which compiler warns about
```c++
if(this->m_extent.prod() != 0u)
{}
```
when extent is signed. GCC 10.2 does not.

Even just removing the `static_cast` may compile without warnings with most compilers, but there was at least one in the CI where it failed, which is why the cast is there. I would guess, that a compiler which warns with `0u` would also warn with `0`, because it does not have the logic see that the sign does not matter when comparing to 0.

I would have to run a changed version through the CI again to be sure and know which compiler cause this line to get a cast, which I am reluctant to do.

jkelling replied:
Ok, the CI ran through on this one, so I will leave it your way. Not sure which compiler it failed on in the past, or if I did this preemptively after all.

bernhardmgruber commented on include/alpaka/mem/buf/oacc/Copy.hpp:
```diff
@@ -0,0 +1,395 @@
+
+                return alpaka::oacc::detail::makeTaskCopyOacc<
+                    alpaka::oacc::detail::TaskCopyOacc,
+                    // oacc::detail::TaskCopyOacc<
+                        TDim,
+                        TViewDst,
+                        TViewSrc,
+                        TExtent>(
+                    //     void(*)(void*,void*,size_t)>(
```
Remove dead comments?

bernhardmgruber commented on include/alpaka/mem/buf/oacc/Copy.hpp:
```diff
@@ -0,0 +1,395 @@
+
+                return alpaka::oacc::detail::makeTaskCopyOacc<
+                    alpaka::oacc::detail::TaskCopyOacc,
+                    // oacc::detail::TaskCopyOacc<
+                        TDim,
+                        TViewDst,
+                        TViewSrc,
+                        TExtent>(
+                    //     void(*)(void*,void*,size_t)>(
```
Dead comments?

bernhardmgruber commented on include/alpaka/mem/buf/oacc/Copy.hpp:
```diff
@@ -0,0 +1,395 @@
+
+                return alpaka::oacc::detail::makeTaskCopyOacc<
+                    alpaka::oacc::detail::TaskCopyOacc,
+                    // oacc::detail::TaskCopyOacc<
+                        TDim,
+                        TViewDst,
+                        TViewSrc,
+                        TExtent>(
+                    //     void(*)(void*,void*,size_t)>(
```
Dead comments?

BenjaminW3 commented on include/alpaka/ctx/block/CtxBlockOacc.hpp:
```diff
@@ -0,0 +1,302 @@
+                int sum;
+// Workaround to use an array in an atomic capture rather than
+// using the data member m_syncCounter array directly.
+// The change is sematically equivlent.
```
```suggestion
// The change is semantically equivalent.
```

jkelling commented:
The latest changes removed `CtxThreadOacc` and merged it into `AccOacc`.

The class `CtxBlockOacc` is only used internally by, now, `AccOacc` and should probably stay this way.
https://github.com/alpaka-group/alpaka/pull/1127/files#diff-2f738a4d0b370767c6efa8a9bbba03d617d297b342f46abbce6825b5993cf54a

@psychocoderHPC @BenjaminW3 @bernhardmgruber should `CtxBlockOacc`
1. stay where it is, in `alpaka/ctx/block`, because it represents a new concept which may come up again, e.g. in SYCL, or
2. be moved into the namespace `oacc::detail` inside `alpaka/acc/AccOacc.hpp`?

bernhardmgruber commented:
> The class CtxBlockOacc is only used internally by, now, AccOacc and should probably stay this way.
> [...]
>
> @psychocoderHPC @BenjaminW3 @bernhardmgruber should `CtxBlockOacc`
> 
> 1. stay where it is, in `alpaka/ctx/block`, because it represents a new concept which may come up again, e.g. in SYCL, or
> 2. be moved into the namespace `oacc::detail` inside `alpaka/acc/AccOacc.hpp`?

My gut feeling tells me that if a thing is internal and only used in one place, then it should be defined in this place. So I am leaning towards moving option 2. But it is not a strong opinion and it can also be done later as a refactoring.

psychocoderHPC commented:
@jkelling Is this PR still WIP or merge ready? Please update the PR headline.

BenjaminW3 commented:
I would like to finish my review but I did not get to it yet.

psychocoderHPC commented on README_OACC.md:
````diff
@@ -0,0 +1,89 @@
+required OpenACC flags for the compiler. Additional flags can be added, e.g:
+- gcc, target x86:
+  ```bash
+    -DCMAKE_CXX_FLAGS="-foffload=disable -fopenacc"
````
only a note: after this PR is merged we should work on adding a CMake option e.g. `ALPAKA_OACC_TARGET` and cover the most used compilers.

jkelling replied:
It actually use the `findPackage(OpenACC)`, -fopenacc is set automaticall (Will fix it alongside sqaushing commits). We could do something for the "-foffload=disable" for GCC.

jkelling commented:
> I would like to finish my review but I did not get to it yet.

I will wait for @BenjaminW3 's review before squashing. When this is done, I will remove the `[WIP]`.

BenjaminW3 commented on include/alpaka/idx/bt/IdxBtOacc.hpp:
```diff
@@ -0,0 +1,128 @@
+        template<
+            typename TDim,
+            typename TIdx>
+        class IdxBtOacc : public concepts::Implements<ConceptIdxBt, IdxBtOacc<TDim, TIdx>>
```
According to the naming we used for OpenMP (`IdxGbLinear`), this should be called `IdxBtLinear`.
I do not see anything OpenAcc specific in here.

BenjaminW3 commented on include/alpaka/ctx/block/CtxBlockOacc.hpp:
```diff
@@ -0,0 +1,302 @@
+        typename TIdx>
+    class CtxBlockOacc final :
+        public WorkDivMembers<TDim, TIdx>,
+        public gb::IdxGbOacc<TDim, TIdx>::BlockShared,
```
We should use `IdxGbLinear` here directly. I do not see any need for the typedef.

BenjaminW3 commented on include/alpaka/ctx/block/CtxBlockOacc.hpp:
```diff
@@ -0,0 +1,302 @@
+        public gb::IdxGbOacc<TDim, TIdx>::BlockShared,
+        public BlockSharedMemDynMember<>,
+        public detail::BlockSharedMemStMemberImpl<4>,
+        public BlockSyncBarrierOacc::BlockShared,
```
I do not see any need for nesting `BlockShared` within `BlockSyncBarrierOacc` here as well. Can't they get merged?

BenjaminW3 commented on include/alpaka/acc/AccOacc.hpp:
```diff
@@ -0,0 +1,514 @@
+        typename TDim,
+        typename TIdx>
+    class AccOacc final :
+        public gb::IdxGbOacc<TDim, TIdx>, // dummy
```
We should not inherit from `IdxGbOacc` if it is not doing something. You should delete `IdxGbOacc` completely.

jkelling replied:
The Idea here was, when `AccOacc` and `CtxThreadOacc` were still separate, to make `AccOacc` formally a valid accelerator which implements all concepts, although the once which can only be used in device code are only actually implemented by `CtxThreadOacc`.

`AccOacc` have been a type only used on the host to reference the backend, but which allows concepts implemented by the backend to be checked by host code. `TaskKernelOacc` in any case creates the on-device Acc instance of type `CtxThreadOacc`. Back then I thougth the Acc might be instantiated by some host code, so I kept it in this double way of having `BlockShared` inside classes which cannot be used on the host. 

This can indeed be removed with the `CtyThreadOacc -> AccOacc` merge in place.

BenjaminW3 commented on include/alpaka/mem/buf/BufOacc.hpp:
```diff
@@ -0,0 +1,650 @@
+            {
+                ALPAKA_DEBUG_MINIMAL_LOG_SCOPE;
+
+                const std::size_t size = static_cast<std::size_t>(extent.prod()) * sizeof(TElem);
```
This should probably be:
```suggestion
                const std::size_t size = static_cast<std::size_t>(extent::getExtentVec(extent).prod()) * sizeof(TElem);
```
Else we can not be sure that we can call `.prod` on it.

BenjaminW3 replied:
Could you apply this to the OMP5 implementation as well please?

jkelling replied:
About the OMP5: do you mean in this PR or in a separate one?

BenjaminW3 replied:
I do not care. Can be this one as well.

jkelling replied:
Included.

BenjaminW3 commented on include/alpaka/mem/buf/BufOacc.hpp:
```diff
@@ -0,0 +1,650 @@
+#if ALPAKA_DEBUG >= ALPAKA_DEBUG_FULL
+                std::cout << __func__ << "alloc'd " << TDim::value
+                    << "D device ptr: " << memPtr << " on device " << dev.m_spDevOaccImpl->iDevice()
+                    << " size " << size << " = " << static_cast<std::size_t>(extent::getExtentVec(extent).prod())*sizeof(TElem) << '\n';
```
Why do you print the size twice?
```suggestion
                    << " size: " << size << '\n';
```

BenjaminW3 replied:
Could you remove calculating the size twice for the OMP5 implementation as well please?

jkelling replied:
Maybe to check if the correct extent vector was returned... . This is very early debug output.

BenjaminW3 commented on include/alpaka/mem/buf/oacc/Set.hpp:
```diff
@@ -0,0 +1,107 @@
+                            pitch
+                        ); // NOP if size is zero
+
+                std::cout << "Set TDim=" << TDim::value << " pitch=" << pitch << " byteExtent=" << byteExtent << std::endl;
```
Please remove or hide in trace #ifdefs

jkelling commented:
@BenjaminW3 if you approve of my changes, I will squash and remove `[WIP}`.

jkelling commented:
@psychocoderHPC please merge. xcode is failing because of CI problems.

bernhardmgruber commented on include/alpaka/acc/AccOacc.hpp:
```diff
@@ -0,0 +1,511 @@
+#ifdef ALPAKA_ACC_ANY_BT_OACC_ENABLED
+
+#if _OPENACC < 201306
+    #error If ALPAKA_ACC_ANY_BT_OACC_ENABLED is set, the compiler has to support OpenACC xx or higher!
```
What is `OpenACC xx`?

jkelling replied:
I never checked which version of OpenACC would be required as a minimum and it is not really possible to check, as no compiler actually works for all tests.

Also technically, no version of OpenACC is sufficient due to the `syncBlockThreads` issue, which requires a barrier implemented based in atomics, which is not allowed by the standard.

bernhardmgruber replied:
I see, that makes sense. Could you maybe add a //TODO that this value needs to be set and why it is not set yet?

jkelling replied:
That should be a doxygen comment then, right? i.e. `//! \todo ...`

jkelling replied:
The value actually checked for amounts to OpenACC 2.0, btw. which may or may not be actually sufficient. At the moment this is more to be a check whether the compiler claims any OpenACC support. I usually used the 2.7 or 3.0 documentation, but checking for these versions would break building with NVHPC, which supports at most 2.6 atm (cannot find documentation on this one now, but I know that it does claim support 2.7).

I could either:
1. Put 2.0 in the error message to make it consistent and have the message tell the truth at leas about the check, or
2. put a `\todo` about which I would not be sure when it could an would be addressed.

I tend to towards 1. now: When something fails because of an actual support issue, the version check will be updated as a bugfix either way.

bernhardmgruber replied:
No doxygen comment. This is a comment for developers that you can grep for or your IDE shows somewhere in a TODO list. I do not see a benefit of making this comment appear in a documentation.

You can then leave the error message as is.

jkelling replied:
I do not know if clang-format can check this, but maybe this should be a point of discussion in https://github.com/alpaka-group/alpaka/pull/1194 . Alpaka contains all kinds of `// TODO` `// \TODO` `//! \TODO` `//@ \todo` ...

jkelling replied:
I changed the message, but without comment. This does not need to be in a todo list. It will come up if and when it get relevant.

bernhardmgruber commented on include/alpaka/acc/AccOacc.hpp:
```diff
@@ -0,0 +1,511 @@
+    #error If ALPAKA_ACC_ANY_BT_OACC_ENABLED is set, the compiler has to support OpenACC xx or higher!
+#endif
+
+// Base classes.
```
Comments like these might be scrambled by clang-format's header sorting. I would remove them.

jkelling replied:
This is also a copy from OpenMP. Could be addressed when clang format is there and a decision has been made on wether they should be there or not also for other backends.

bernhardmgruber commented on include/alpaka/atomic/AtomicOaccBuiltIn.hpp:
```diff
@@ -0,0 +1,388 @@
+    class AtomicOaccBuiltIn
+    {
+    public:
+        //-----------------------------------------------------------------------------
```
Why do we have these lines here?

jkelling replied:
Copy from OpenMP.

bernhardmgruber replied:
That is not a reason for keeping them. But do as you want.

jkelling replied:
The OpenMP atomics file has the same lines in the same place, this minimizes the diff to OpenMP.

bernhardmgruber commented on include/alpaka/atomic/AtomicOaccBuiltIn.hpp:
```diff
@@ -0,0 +1,388 @@
+                T old;
+                auto & ref(*addr);
+                // atomically update ref, but capture the original value in old
+#if !BOOST_COMP_PGI || defined TPR28628 // triggers PGI TPR28628, not atomic until fixed
+                #pragma acc atomic capture
+#endif
```
Do I understand correctly, that there is a compiler with a specific version where `alpaka::atomicExch` will silently not be atomic?
I think this should be at least a warning or even a hard error.

jkelling replied:
We need this to compile work in an experimental PIConGPU version. This whole backend is experimental because there is no compiler support. The affected compiler version cannot be considered production-ready anyway. Once there is a working compiler, the old compiler version can just be declared incompatible altogether.

bernhardmgruber replied:
Well, if this PR is merged, this version becomes part of a public alpaka release and people are going to download and try it. I would be discouraged as a client if that caused a bug in my code, even if it's in 2 years from.

Could you please add something like a `#pragma message("Atomic exchange will not be atomic. Sorry :/")`?

jkelling replied:
Ok. Why not #warning ?

bernhardmgruber replied:
Because #warning is not supported by MSVC and will give you a compile error. Which would also be fine in this case :D #pargma message should work on MSVC, GCC and clang, and if not supported, must be ignored.

jkelling replied:
In this case, we haven an issue here: https://github.com/alpaka-group/alpaka/blob/develop/include/alpaka/example/ExampleDefaultAcc.hpp#L48

Will this message be a warning and be elevated to an error by `-Werror`?

bernhardmgruber replied:
This is a hard error on MSVC and I was too lazy yet to open an issue.

bernhardmgruber commented on include/alpaka/block/sync/BlockSyncBarrierOacc.hpp:
```diff
@@ -0,0 +1,48 @@
+        std::uint8_t mutable m_generation = 0u;
+        // NVHPC 20.7: initializer causes warning:
+        // NVC++-W-0155-External and Static variables are not supported in acc routine - _T139951818207704_37530
+        int mutable m_syncCounter[4] {0,0,0,0};
```
Where does the number 4 come from?

jkelling replied:
Space for two generation times two slots in https://github.com/jkelling/alpaka/blob/openacc/include/alpaka/ctx/block/CtxBlockOacc.hpp#L105 .

bernhardmgruber replied:
Thx! Could you please add a comment for that or move 4 into a named constant?

jkelling replied:
A constant would suggest safety which is not there, also: I cannot come up with a descriptive name. So, a comment it is.

bernhardmgruber commented on include/alpaka/block/sync/BlockSyncBarrierOacc.hpp:
```diff
@@ -0,0 +1,48 @@
+        // NVHPC 20.7: initializer causes warning:
+        // NVC++-W-0155-External and Static variables are not supported in acc routine - _T139951818207704_37530
+        int mutable m_syncCounter[4] {0,0,0,0};
+        int mutable m_result;
```
Is it a performance degradation if we initialize this value? Especially because all other members have default initializers.

jkelling replied:
It is initialized before each use (by https://github.com/jkelling/alpaka/blob/openacc/include/alpaka/ctx/block/CtxBlockOacc.hpp#L214 ). After each us, it should be considered uninitalized again, so I do not see why it should be initialized even from a semantics standpoint.

bernhardmgruber replied:
I know it's initialized later. But the question is: could this type be safer by having all members default initialized on construction without a performance penalty? If the `int mutable m_result = 0;` does not hurt, I would add it.

jkelling replied:
Variables should not be initialized at all cost, because it effectively mutes use-before-init compiler warnings. If I do not actually want to rely on it having this value before first using it, I would consider initialization bad style.

jkelling replied:
Btw. I do not know if there would be runtime cost, there might be on a GPU. This is more of a style /semantics issue for me.

bernhardmgruber replied:
I would argue the store to this variable will be eliminated by a the compiler because it can see the store afterwards with the value you need in your algorithm later. This is not a style issue, but a safety issue. That's why it is also in the core guidelines: https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Pro-type-memberinit

jkelling replied:
Eliminating the store affects only the performance aspect.

Initializing it in the ctor is more likely to hide a bug: This variable is reused by subsequent invocations of that algorithm. If one forgot to init it at the beginning of the function, you would have a bug.

I usually agree with initializing things in ctors, that is what they are for, but this is the exception to the rule.

BenjaminW3 replied:
I am with @jkelling here.

bernhardmgruber commented on include/alpaka/dev/DevOacc.hpp:
```diff
@@ -0,0 +1,341 @@
+                ALPAKA_FN_HOST auto getAllExistingQueues() const
+                -> std::vector<std::shared_ptr<IGenericThreadsQueue<DevOacc>>>
+                {
+                    std::vector<std::shared_ptr<IGenericThreadsQueue<DevOacc>>> vspQueues;
+
+                    std::lock_guard<std::mutex> lk(m_Mutex);
+                    vspQueues.reserve(m_queues.size());
```
```suggestion
                    std::lock_guard<std::mutex> lk(m_Mutex);
                    std::vector<std::shared_ptr<IGenericThreadsQueue<DevOacc>>> vspQueues(m_queues.size());
```

jkelling replied:
`reserve()` is leaving the size of the vector 0, unlike `resize()`, which is what the ctor call amounts to.

bernhardmgruber replied:
Oh sorry, my mistake. Good catch!

bernhardmgruber commented on include/alpaka/dev/DevOacc.hpp:
```diff
@@ -0,0 +1,341 @@
+        //-----------------------------------------------------------------------------
+        auto operator=(DevOacc &&) -> DevOacc & = default;
+        //-----------------------------------------------------------------------------
+        ALPAKA_FN_HOST auto operator==(DevOacc const & rhs) const
```
Just because it noticed it here. What is the rational for explicitely marking a function as `ALPAKA_FN_HOST`? We could just omit that and it would have the same meaning and be shorter?

jkelling replied:
It is true, that `ALPAKA_FN_HOST` is just empty for OpenACC. This is intended in part documentation ("do not try using this in device code, which we cannot check") and mostly for consistency with other backend, where `ALPAKA_FN_HOST` actually does something  (i.e. CUDA).

bernhardmgruber replied:
But even for CUDA, would it make a difference in semantic? AFAIK you cannot call an unannotated function from a device function as well.

BenjaminW3 commented:
I noticed it too late, but shouldn't we update the tables in the ReadMe for both, OpenACC and OpenMP 5?

SimeonEhrig commented:
> I noticed it too late, but shouldn't we update the tables in the ReadMe for both, OpenACC and OpenMP 5?

In general, maybe we can integrate the `README_OACC.md` and `README_OMP5.md` in the official sphinx doc documentation. We have a sections for [cmake arguments](https://alpaka.readthedocs.io/en/latest/advanced/cmake.html) and [internal, backend specific develop stuff](https://alpaka.readthedocs.io/en/latest/dev/backends.html).

jkelling commented:
> I noticed it too late, but shouldn't we update the tables in the ReadMe for both, OpenACC and OpenMP 5?

Right. but since we missed this for the OpenMP5 PR too, it seems fitting to do this in a separate PR for both.
